Bump bounds

maralorn opened the pull request:
This bumps a number of bounds for stackage-nightly compatibility.

Needs to make tiny changes for lens 5.0 compat.

maralorn commented:
Actually I think that the lens lower bound could probably be loosened if necessary because itraversed might be optional.

Ericson2314 commented on src/Data/Patch/IntMap.hs:
```diff
@@ -52,7 +52,6 @@ instance FunctorWithIndex Int PatchIntMap
 instance FoldableWithIndex Int PatchIntMap
 instance TraversableWithIndex Int PatchIntMap where
   itraverse = itraversed . Indexed
-  itraversed = _Wrapped .> itraversed <. traversed
```
Isn't this going to blow up?

maralorn replied:
I don‘t think so, the class was changed so that it only needs `itraverse` and lens implements `itraversed` based on it. 
But apparently we need CPP for lens-4 compat, if that’s desired.

https://hackage.haskell.org/package/lens-5.1/changelog:

> The FunctorWithIndex, FoldableWithIndex and TraversableWithIndex type classes have been migrated to a new package, [indexed-traversable](https://hackage.haskell.org/package/indexed-traversable).
>
>The imapped, ifolded and itraversed methods are now top-level functions. If you are not defining these methods in your instances, you don't need to change your definitions.
>
>Beware: the optics-core package (versions <0.4) defines similar classes, and will also migrate to use indexed-traversable classes. Therefore, you might get duplicate instance errors if your package defines both.
>
>If you define your own FunctorWithIndex etc. instances, we recommend that you depend directly on the indexed-traversable package. If you want to continue support lens-4 users, you may write
>
> ```
> -- from indexed-traversable
> import Data.Functor.WithIndex
>
> -- from lens
> import qualified Control.Lens as L
>
> -- your (indexed) container
> data MySeq a = ...
>
> -- indexed-traversable instance
> instance FunctorWithIndex     Int MySeq where imap = ...
> instance FoldableWithIndex    Int MySeq where ifoldMap = ...
> instance TraversableWithIndex Int MySeq where itraverse = ...
> 
> -- lens <5 instance, note the !
> #if !MIN_VERSION_lens(5,0,0)
> instance L.FunctorWithIndex     Int MySeq where imap = imap
> instance L.FoldableWithIndex    Int MySeq where ifoldMap = ifoldMap
> instance L.TraversableWithIndex Int MySeq where itraverse = itraverse
> #endif
>```
>
> In other words, always provide indexed-traversable instances. If your package depends on lens and allows lens-4, you should additionally provide instances for lens-4 type classes that can reuse the indexed-traversable instances.

Ericson2314 replied:
Yeah but see the definition in `lens` of `itraversed` is now:
```haskell
itraversed = conjoined traverse (itraverse . indexed)
```
but `patch` has
```haskell
itraverse = itraversed . Indexed
```
here. I think that will mean divergence.

maralorn replied:
Oh, sorry. I hadn’t actually looked into the other definition.

maralorn committed:
Bump bounds for base, hlint and semialign

maralorn committed:
Make compatible with lens 5

This follows the migration guide for the *WithIndex classes in
https://hackage.haskell.org/package/lens-5.1/changelog

maralorn commented:
Okay I tried to solve this proper. I didn‘t really understand what those functions do, but by equational reasoning this should work and not loop.

I have run the cabal test with lens < 5 and lens >= 5.

Ericson2314 commented:
Thanks!
